File: apps/web/tests/contracts/printingRecentChanges.contract.test.js
// Contract tests for printing recent-change overlay lookup.
//
// These assert the shared fixtures in packages/contracts, which the Dart
// implementation in apps/mobile is held to as well. A failure here means either
// this file's logic drifted from mobile, or the agreed behaviour changed and both
// sides plus the fixture need updating. See packages/contracts/README.md.
import contract from '../../../../packages/contracts/printing_recent_changes.json';
import { lookupPrintingRecentChanges } from '../../src/utils/recentMovers.js';

describe('printing recent changes contract', () => {
    const today = contract.today;

    describe('lookupPrintingRecentChanges', () => {
        contract.lookup_cases.forEach((c) => {
            it(c.name, () => {
                const overlays = lookupPrintingRecentChanges({
                    today,
                    source: c.source,
                    snapshotsByCard: c.snapshots_by_card,
                    currentLows: c.current_lows,
                    cardIds: c.card_ids,
                });
                expect(overlays).toEqual(c.expected);
                for (const row of overlays) {
                    expect(row.percent_change).not.toBe(0);
                    expect(row.amount_change).toBeCloseTo(row.latest_low - row.start_low);
                    expect(row.percent_change).toBeCloseTo(
                        (row.latest_low - row.start_low) / row.start_low,
                    );
                }
            });
        });
    });

    it('includes an under-floor cheap spike', () => {
        const cheap = contract.lookup_cases.find((c) => c.name.includes('0.30'));
        const overlays = lookupPrintingRecentChanges({
            today,
            source: cheap.source,
            snapshotsByCard: cheap.snapshots_by_card,
            currentLows: cheap.current_lows,
            cardIds: cheap.card_ids,
        });
        expect(overlays).toHaveLength(1);
        expect(overlays[0].card_id).toBe('cheap-spike');
        expect(overlays[0].start_low).toBeLessThan(1);
    });

    it('omits 0%, missing window, and outliers', () => {
        const unchanged = contract.lookup_cases.find((c) => c.name.includes('2.00'));
        expect(lookupPrintingRecentChanges({
            today,
            source: unchanged.source,
            snapshotsByCard: unchanged.snapshots_by_card,
            currentLows: unchanged.current_lows,
            cardIds: unchanged.card_ids,
        })).toEqual([]);

        const window = contract.lookup_cases.find((c) => c.name.includes('2-day'));
        expect(lookupPrintingRecentChanges({
            today,
            source: window.source,
            snapshotsByCard: window.snapshots_by_card,
            currentLows: window.current_lows,
            cardIds: window.card_ids,
        }).map((r) => r.card_id)).toEqual(['in-window']);

        const outlier = contract.lookup_cases.find((c) => c.name.includes('outlier'));
        expect(lookupPrintingRecentChanges({
            today,
            source: outlier.source,
            snapshotsByCard: outlier.snapshots_by_card,
            currentLows: outlier.current_lows,
            cardIds: outlier.card_ids,
        }).map((r) => r.card_id)).toEqual(['at-percent-cap']);
    });

    it('never coerces a null Low to 0', () => {
        const nullCase = contract.lookup_cases.find((c) => c.name.includes('null Low'));
        expect(nullCase.snapshots_by_card['null-only'][0].tcg_low).toBeNull();
        const overlays = lookupPrintingRecentChanges({
            today,
            source: nullCase.source,
            snapshotsByCard: nullCase.snapshots_by_card,
            currentLows: nullCase.current_lows,
            cardIds: nullCase.card_ids,
        });
        expect(overlays.map((r) => r.card_id)).not.toContain('null-only');
        expect(overlays.some((r) => r.start_low === 0 || r.latest_low === 0)).toBe(false);
    });
});

File: apps/mobile/test/contracts/printing_recent_changes_contract_test.dart
// Contract tests for printing recent-change overlay lookup.
//
// These assert the shared fixtures in packages/contracts, which the JavaScript
// implementation in apps/web is held to as well. A failure here means either this
// package's logic drifted from web, or the agreed behaviour changed and both sides
// plus the fixture need updating. See packages/contracts/README.md.
import 'package:fabtrades/core/logic/recent_movers.dart';
import 'package:flutter_test/flutter_test.dart';

import 'contract_fixtures.dart';

void main() {
  final contract = loadContract('printing_recent_changes');
  final today = contract['today'] as String;

  Map<String, List<Map<String, dynamic>>> snapshots(
    Map<String, dynamic> c,
  ) {
    final raw = Map<String, dynamic>.from(c['snapshots_by_card'] as Map);
    return raw.map(
      (id, snaps) => MapEntry(
        id,
        (snaps as List)
            .map((e) => Map<String, dynamic>.from(e as Map))
            .toList(),
      ),
    );
  }

  Map<String, Map<String, dynamic>> currentLows(Map<String, dynamic> c) {
    final raw = Map<String, dynamic>.from(c['current_lows'] as Map);
    return raw.map(
      (id, row) => MapEntry(id, Map<String, dynamic>.from(row as Map)),
    );
  }

  List<String> cardIds(Map<String, dynamic> c) =>
      (c['card_ids'] as List).map((e) => e as String).toList();

  group('lookupPrintingRecentChanges', () {
    for (final c in contractCases(contract, 'lookup_cases')) {
      final name = c['name'] as String;
      test(name, () {
        final overlays = lookupPrintingRecentChanges(
          today: today,
          source: c['source'] as String,
          snapshotsByCard: snapshots(c),
          currentLows: currentLows(c),
          cardIds: cardIds(c),
        );
        expect(overlays.map((r) => r.toJson()).toList(), c['expected']);
        for (final row in overlays) {
          expect(row.percentChange, isNot(0));
          expect(
            row.amountChange,
            closeTo(row.latestLow - row.startLow, 1e-9),
          );
          expect(
            row.percentChange,
            closeTo((row.latestLow - row.startLow) / row.startLow, 1e-9),
          );
        }
      });
    }
  });

  test('under-floor cheap spike is included', () {
    final cheap = contractCases(contract, 'lookup_cases').firstWhere(
      (c) => (c['name'] as String).contains('0.30'),
    );
    final overlays = lookupPrintingRecentChanges(
      today: today,
      source: cheap['source'] as String,
      snapshotsByCard: snapshots(cheap),
      currentLows: currentLows(cheap),
      cardIds: cardIds(cheap),
    );
    expect(overlays, isNotEmpty);
    expect(overlays.single.cardId, 'cheap-spike');
    expect(overlays.single.startLow, lessThan(1));
  });

  test('0%, missing window, and outliers are omitted', () {
    for (final needle in ['2.00', '2-day', 'outlier']) {
      final c = contractCases(contract, 'lookup_cases').firstWhere(
        (row) => (row['name'] as String).contains(needle),
      );
      final overlays = lookupPrintingRecentChanges(
        today: today,
        source: c['source'] as String,
        snapshotsByCard: snapshots(c),
        currentLows: currentLows(c),
        cardIds: cardIds(c),
      );
      if (needle == '2.00') {
        expect(overlays, isEmpty);
      } else if (needle == '2-day') {
        expect(overlays.map((r) => r.cardId), ['in-window']);
      } else {
        expect(overlays.map((r) => r.cardId), ['at-percent-cap']);
      }
    }
  });

  test('null Low is never coerced to 0', () {
    final nullCase = contractCases(contract, 'lookup_cases').firstWhere(
      (c) => (c['name'] as String).contains('null Low'),
    );
    final snaps = snapshots(nullCase)['null-only']!;
    expect(snaps.first['tcg_low'], isNull);
    final overlays = lookupPrintingRecentChanges(
      today: today,
      source: nullCase['source'] as String,
      snapshotsByCard: snapshots(nullCase),
      currentLows: currentLows(nullCase),
      cardIds: cardIds(nullCase),
    );
    expect(overlays.map((r) => r.cardId), isNot(contains('null-only')));
    expect(overlays.any((r) => r.startLow == 0), isFalse);
    expect(overlays.any((r) => r.latestLow == 0), isFalse);
  });
}
